### PR TITLE
fix #670 (Copy pasting value adds incorrect additional comma/decimal to rawValue), add e2e tests related to this fix

### LINKED ISSUE #670: Copy pasting value adds incorrect additional comma/decimal to rawValue
If I copy paste a number such as "123456" in an Autonumeric field, the rawValue seems to add an additional comma to the end of the number, resulting in an incorrect number: `123456.00,`

I can reproduce this on the example on the Autonumeric homepage:

afbeelding

But for some strange reason sometimes it *does* work correctly. See this gif of me continuously copy pasting "123456" in the input field (all with the french settings)

copypaste

My browser locale is Dutch. We've had several errors in production with Polish locale (and Autonumeric configured for polish, which also has the currency symbol as a suffix similar to the French setting)

I was not able to reproduce it with the northAmerican preset. So I would assume it has something to do with either the currency symbol placement, or the decimal character (it being a comma in Dutch and Polish)

Hope you can help with this weird behavior.

### PR DESCRIPTION
**Pull request by kferenc617**
See #670 for further details

### CODE REVIEW

**Comment by AlexandreBonneau** [comment]
Thanks @kferenc617 !
Do you think it would be possible to update the e2e tests to use a new input in `index.html` instead of reusing the one created for testing issue 387? That would be more consistent with the previous issue fixes.

**Comment by kferenc617** [comment]
@AlexandreBonneau: sure, no problem. I was also thinking about this option. I will post a commit soon.

**Comment by kferenc617** [comment]
done (I also found another branch in _onPaste (onInvalidPaste.clamp) that needs fixing and added an e2e test for it

**Comment by kferenc617** [comment]
@AlexandreBonneau: I have a feeling that the onInvalidPaste.ignore is also affected, please wait a little bit with the merging. I will investigate it soon and come back to you with the results.

**Comment by kferenc617** [comment]
I've reviewed the onInvalidPaste=ignore case and similarly it should also be fixed, because when the pasted value is outside of the range the value should not be changed (according to the comment in _onPaste() and the documentation also agrees on it (https://autonumeric.org/guide#onInvalidPaste).

I've committed the fix and I've added a e2e "sub-test" for this case. (without the fix, the new test fails, after the fix, all unit and e2e-tests succeed)

@AlexandreBonneau: I was wondering if it would be cleaner to move this piece of code (this.formatted = true) into the beginning of the _onPaste() function instead of repeating it in each and every branches because this flag should always be set to true in this function (when set() succeeds and also when the function throws an exception). What do you think?

File: src/AutoNumericHelper.js [hunk 1]
```diff
@@ -481,7 +481,7 @@ export default class AutoNumericHelper {
      */
     static isSeleniumBot() {
         // noinspection JSUnresolvedVariable
-        return window.navigator.webdriver === true;
+        return window.navigator.webdriver === true && this.browser().name === 'firefox';  // Should return false if executed under chromium/webdriver (both the geckodriver and the webdriver set window.navigator.webdriver to true)
```
Comment (AlexandreBonneau) [inline src/AutoNumericHelper.js:485]: I'm not sure why there is a test for Firefox. On Chrome this `isSeleniumBot()` function will never be true, isn't?

**Comment by AlexandreBonneau** [comment]
> @AlexandreBonneau: I was wondering if it would be cleaner to move this piece of code (this.formatted = true) into the beginning of the _onPaste() function instead of repeating it in each and every branches because this flag should always be set to true in this function (when set() succeeds and also when the function throws an exception). What do you think?

Indeed I do not like repeating the code and try to DRY it as much as possible (hence the comment on line 7375).

However managing the paste event was a big headache back then, and to be honest it's been a while since I touched this part of the code. I remember that there was a LOT of tiny use cases where all the ifs and thens made sense to get a coherent UX.

That said, I do not like using `this.formatted = true` which in essence means that the value has been formatted (which it is not when used there), just to prevent the keyup event on the 'v' key.
I much prefer being explicit with variable names and usages, and would prefer to use a specific variable `preventKeyup = true` in its place.

What are your thoughts about that?

File: src/AutoNumericHelper.js (hunk 1 above)
Comment (kferenc617) [reply src/AutoNumericHelper.js:485]: It seems that isSeleniumBot was created to handle an issue with e2e tests run under Firefox/geckodriver. This function has been called only from AutoNumericHelper.character() to handle a bug in geckodriver. Running e2e tests as of today, this function returns false and the workaround in AutoNumericHelper.character() is not needed anymore (if isSeleniumBot returns true, a workaround is activated in character()). 

Actually, the extra condition (in my PR) was needed because this function returned true on Chrome, but the workaround is not needed on this browser.

If it is not feasible that e2e tests will be run under firefox/geckodriver again in the foreseeable future (considering that firefox's market share declines - unfortunately), I could delete isSeleniumBot() and the workaround branch in character(). 

What do you think?

**Comment by kferenc617** [comment]
Regarding the this.formatted flag: I've been looking at the code for a while and not convinced in either way. I had the impression that originally it was created to prevent double formatting of input values (e.g.: see the comment at the beginning of _onKeydown()), but its use and meaning may have already been shifted and such a change is better be reflected in the name. 

The flag affects execution in _onKeyup(), but also in _checkPaste(). 

Probably _checkPaste() is not needed anymore, which would simplify things. I'll test paste functionality more to determine if there is an edge case where _checkPaste() is needed, but if event order is always _onKeyDown -> _onPaste() -> _onKeyUp(), then _checkPaste() is not needed, because this.formatted will be true after _onPaste().

Do you have information if event order is different (e.g.: on another platform than Windows, or on a different browser than FF or Chromium)?

**Comment by kferenc617** [comment]
I've tested manually in Chrome/Firefox on Windows, and when the above fix is in effect, _checkPaste() is not needed. I've also tested using the paste-related e2e tests and this function (or more precisely its internal parts where the input value is modified) is not executed at all. 

Basically it is expected, because _checkPaste() returns early if this.formatted is true, and since _onPaste() sets this.formatted to true in all branches, this function is not needed anymore. 

Actually I think it was this piece of code that may have caused some problems with the rawValue/display value.
Once _checkPaste() is removed, most probably this.valuePartsBeforePaste is not needed either.

What if this.formatted is renamed to this.preventFormattingInKeyup? 

Would you be interested in a commit that contains this flag renamed and _checkPaste() and this.valuePartsBeforePaste removed?

**Comment by AlexandreBonneau** [comment]
If indeed your findings show that _checkPaste() is not used anymore in the end to end tests, then it might be time to retire it.
Any effort in streamlining the input chain events and paste actions is welcome.

Have you tried to paste numbers mixed with strings, for instance '123milimeters' (which should use 123)?

**Comment by kferenc617** [comment]
I've investigated and _checkPaste() is not needed anymore. I've also removed valuePartsBeforePaste field, because it was needed by _checkPaste(). I've also added a tests case when '123millimeters' is pasted. All unit and e2e tests succeed.

File: src/AutoNumericHelper.js (hunk 1 above)
Comment (AlexandreBonneau) [reply src/AutoNumericHelper.js:485]: Letting Chrome take all the marketshare is like shooting ourselves in the foot. I'm still haunted with IE days, their monopolistic position and their shitty support for standards.
Firefox is and will stay for the foreseeable future my primary target when developing web-related stuff, especially since Mozilla tend to respect the standards quite a bit more than Google.

Comment (kferenc617) [reply src/AutoNumericHelper.js:485]: Actually, I also use FF for both daily use and development - debugging web stuff. Sometimes I wonder if Mozilla's reluctance to implement new/shiny web standards accounts for the fewer high scored vulnerabilities in it.
